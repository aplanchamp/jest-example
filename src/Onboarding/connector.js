export const updateOnboarding = onboarding => {
  console.log("updateOnboarding", onboarding);
};
