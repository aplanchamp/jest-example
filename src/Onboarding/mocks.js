export const startupNotification = {
  currentPage: {
    displayTemplate: "slideShow",
    BOName: "On Boarding démarrage FREE",
    BOLayoutName: "[FREE] On Boarding démarrage",
    path: "/slideshows/pid103110-on-boarding-demarrage-free.html"
  },
  tracking: {
    omniture: {
      prop4: "Slideshows",
      prop5: "Slideshows - On Boarding demarrage FREE",
      pageName: "Slideshows - On Boarding demarrage FREE",
      channel: "myCMob",
      eVar31: "[FREE] On Boarding démarrage",
      prop10: "Contenu",
      prop11: "Autres",
      prop12: "Consumer"
    }
  },
  slides: [
    {
      URLImageRegularLandscape:
        "https://thumb.canalplus.pro/bran/unsafe/{resolutionXY}/image/44/1/1_onboarding.59441.jpg",
      buttons: []
    },
    {
      URLImageRegularLandscape:
        "https://thumb.canalplus.pro/bran/unsafe/{resolutionXY}/image/44/2/2_onboarding.59442.jpg",
      buttons: []
    },
    {
      URLImageRegularLandscape:
        "https://thumb.canalplus.pro/bran/unsafe/{resolutionXY}/image/44/4/3_onboarding.59444.jpg",
      buttons: []
    },
    {
      URLImageRegularLandscape:
        "https://thumb.canalplus.pro/bran/unsafe/{resolutionXY}/image/44/3/4_onboarding.59443.jpg",
      buttons: []
    }
  ]
};
