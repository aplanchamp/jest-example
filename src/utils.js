export const replace = url => {
  console.log(url);
};
