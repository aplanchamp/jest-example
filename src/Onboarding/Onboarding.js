import React, { Component } from "react";

import { replace } from "../utils";

class Onboarding extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentSlideIndex: 0
    };
  }

  componentDidMount() {
    if (this.props.onboardingShouldNotBeDisplayed) {
      replace("/home");
    }
  }

  onLeaveOnboarding = () => {
    this.props.updateOnboarding({ alreadyDisplayed: true });
  };

  next = () => {
    this.setState({ currentSlideIndex: this.state.currentSlideIndex + 1 });
  };

  render() {
    const { startupNotification } = this.props;
    const currentPage = startupNotification && startupNotification.currentPage;
    return currentPage
      ? this.renderTemplateComponent(
          currentPage.displayTemplate,
          startupNotification
        )
      : null;
  }

  renderTemplateComponent(displayTemplate, { slides }) {
    const currentSlide = slides[this.state.currentSlideIndex];
    const currentSlideImgSrc =
      currentSlide && currentSlide.URLImageRegularLandscape
        ? currentSlide.URLImageRegularLandscape.replace(
            "{resolutionXY}",
            "450x200"
          )
        : "";
    switch (displayTemplate) {
      case "slideShow":
        return (
          <div>
            <img
              className="slideShow-image"
              src={currentSlideImgSrc}
              alt="onboarding-img"
            />
            <button onClick={this.next}>Next</button>
            <button onClick={this.onLeaveOnboarding}>Leave</button>
          </div>
        );

      default:
        return null;
    }
  }
}

export default Onboarding;
