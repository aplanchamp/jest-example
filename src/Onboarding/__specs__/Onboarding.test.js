import React from "react";
import Onboarding from "../Onboarding";
import ReactTestRenderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import { startupNotification } from "../mocks";

import * as utils from "../../utils";

describe("Onboarding component", () => {
  describe("Snapshots", () => {
    it("should render snapshot - react-test-renderer", () => {
      const renderer = ReactTestRenderer.create(
        <Onboarding startupNotification={startupNotification} />
      );
      expect(renderer.toJSON()).toMatchSnapshot();
    });

    it("should render snapshot - @testing-library/react", () => {
      const { asFragment } = render(
        <Onboarding startupNotification={startupNotification} />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render null - react-test-renderer", () => {
      const renderer = ReactTestRenderer.create(<Onboarding />);
      expect(renderer.toJSON()).toMatchSnapshot();
    });

    it("should render null  - @testing-library/react", () => {
      const { asFragment } = render(<Onboarding />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("should redirect to home", () => {
    const spy = jest.spyOn(utils, "replace");
    render(<Onboarding onboardingShouldNotBeDisplayed />);
    expect(spy).toHaveBeenCalledWith("/home");
  });

  it("should call the updateOnboarding prop if user clicks the 'Leave' button", () => {
    const updateOnboarding = jest.fn();
    const test = render(
      <Onboarding
        startupNotification={startupNotification}
        updateOnboarding={updateOnboarding}
      />
    );
    console.log(test);
    fireEvent.click(test.getByText(/Leave/i));
    expect(updateOnboarding).toHaveBeenCalledWith({
      alreadyDisplayed: true
    });
  });

  it("should update the displayed image if the user clicks the 'Next' button", () => {
    const { getByText, getByAltText } = render(
      <Onboarding startupNotification={startupNotification} />
    );
    const imgNode = getByAltText("onboarding-img");
    fireEvent.click(getByText(/Next/i));

    expect(imgNode.src).toEqual(
      "https://thumb.canalplus.pro/bran/unsafe/450x200/image/44/2/2_onboarding.59442.jpg"
    );
  });
});
