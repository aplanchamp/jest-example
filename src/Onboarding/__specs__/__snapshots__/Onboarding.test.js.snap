// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboarding component Snapshots should render null  - @testing-library/react 1`] = `<DocumentFragment />`;

exports[`Onboarding component Snapshots should render null - react-test-renderer 1`] = `null`;

exports[`Onboarding component Snapshots should render snapshot - @testing-library/react 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="onboarding-img"
      class="slideShow-image"
      src="https://thumb.canalplus.pro/bran/unsafe/450x200/image/44/1/1_onboarding.59441.jpg"
    />
    <button>
      Next
    </button>
    <button>
      Leave
    </button>
  </div>
</DocumentFragment>
`;

exports[`Onboarding component Snapshots should render snapshot - react-test-renderer 1`] = `
<div>
  <img
    alt="onboarding-img"
    className="slideShow-image"
    src="https://thumb.canalplus.pro/bran/unsafe/450x200/image/44/1/1_onboarding.59441.jpg"
  />
  <button
    onClick={[Function]}
  >
    Next
  </button>
  <button
    onClick={[Function]}
  >
    Leave
  </button>
</div>
`;
